Lib/asyncio/runners.py

```diff
@@ -136,7 +134,11 @@ def _lazy_init(self):
             return
         if self._loop_factory is None:
             self._loop = events.new_event_loop()
-            self._set_event_loop = True
+            if not self._set_event_loop:
+                # Call set_event_loop only once to avoid calling
+                # attach_loop multiple times on child watchers
+                events.set_event_loop(self._loop)
+                self._set_event_loop = True
```

Comment:
I have some questions.

First, why is `self._set_event_loop` not set, and `events.set_event_loop()` not called, when `self._loop_factory` is not `None`? The [docs](https://docs.python.org/3.11/library/asyncio-runner.html#runner-context-manager) do not imply any semantic difference -- they just imply it defaults to `new_event_loop()`. (This is not something you changed, but I noticed this while trying to understand the code -- is this an oversight or is there a reason that just didn't get documented? FWIW When I add this on macOS, one test fails: `test_create_subprocess_fails_with_inactive_watcher`.)

Second, you are moving the call to `set_event_loop()` to an earlier stage. The original code did this after the task is created and the `SIGINT` handler is installed (and inside the `try/finally` that uninstalls it). Your new code does it sooner. Is that better? Worse? Do you have an opinion?

Finally, the name `_set_event_loop` is a bit ambiguous -- is it meant to indicate that the event loop *was* set, or does it mean that the event loop *should* be set? In the original code it kind of waffles between the two -- setting it causes the loop to be set in `run()` and also causes it to be reset in `close()`. In your code we can interpret it as always meaning that it *was* set, and must be reset in `close()`. But then a different name would be better.

Reply:
> This is not something you changed, but I noticed this while trying to understand the code -- is this an oversight or is there a reason that just didn't get documented?

This was changed in the original "Runner doesn't call set_event_loop issue"

Reply:
Interesting, I didn't realize that. A resolved comment in #94593 says:

> Well currently if loop factory was not supplied then it calls events.new_event_loop() which initializes the policy registry hence we must call set_event_loop. However if the user provides a loop factory then we don't have to set the event loop.

This confuses me. What does "policy registry" refer to? Note that `BaseDefaultEventLoopPolicy.new_event_loop()` states in its docstring:

> You must call set_event_loop() to make this the current event loop.

@kumaraditya303 I am very confused now.

Reply:
(I realize I got this backwards -- you are correctly calling `set_event_loop()` when you're calling `new_event_loop()`. But where is it written that if the user supplies a `loop_factory` function, that function must both create and set the event loop?

Reply:
> that function must both create and set the event loop?

The loop_factory function doesn't actually need to call set_event_loop because new code doesn't need it called, only old code requires the sensitive call to `attach_loop`

Reply:
So you are equating "old code" with "doesn't specify loop_factory". But the docs don't bear that out (they just imply that the default loop factory is `new_event_loop`). Are the docs just missing this? Or is there something else I'm missing?

Reply:
> Are the docs just missing this? Or is there something else I'm missing?

Yeah docs are missing the information

https://github.com/python/cpython/pull/94593#issuecomment-1175941729

However the information needed there is subtle: `set_event_loop` is for use with the deprecated `get_event_loop` and the "not deprecated by omission" `get_event_loop_policy().get_event_loop()`

There's an additional set of posix only trouble with the child watchers where `set_event_loop` calls `attach_loop`, the recommendation here appears to be for asyncio users to just not touch the child watchers at all and use the ThreadedChildWatcher which doesn't need the global event loop.

Which is why the documentation should not be: 
"that function must both create and set the event loop"
but instead describe why the function should not call `set_event_loop`, however it's difficult to write the documentation without the deprecations in place

Reply:
> First, why is self._set_event_loop not set, and events.set_event_loop() not called, when self._loop_factory is not None? The [docs](https://docs.python.org/3.11/library/asyncio-runner.html#runner-context-manager) do not imply any semantic difference -- they just imply it defaults to new_event_loop(). (This is not something you changed, but I noticed this while trying to understand the code -- is this an oversight or is there a reason that just didn't get documented? 

`loop_factory` is provided by user when they want to use their own custom loop factory. If you provide your own loop factory then, you are supposed to manage and set it as the current in the policy registry if you have one. `new_event_loop` requires `set_event_loop` to be called hence we are calling it here.


> FWIW When I add this on macOS, one test fails: test_create_subprocess_fails_with_inactive_watcher.)


Can you elaborate on that? The CI is green and all tests have passed on macOS. Can you provide a stack trace?


> Second, you are moving the call to set_event_loop() to an earlier stage. The original code did this after the task is created and the SIGINT handler is installed (and inside the try/finally that uninstalls it). Your new code does it sooner. Is that better? Worse? Do you have an opinion?


It is better to set the event loop earlier since we want to reflect that we have created a loop instance from a policy factory and it is active.


> Finally, the name _set_event_loop is a bit ambiguous -- is it meant to indicate that the event loop was set, or does it mean that the event loop should be set? In the original code it kind of waffles between the two -- setting it causes the loop to be set in run() and also causes it to be reset in close(). In your code we can interpret it as always meaning that it was set, and must be reset in close(). But then a different name would be better.


Suggestion for better name is welcome!

Reply:
> This confuses me. What does "policy registry" refer to? Note that BaseDefaultEventLoopPolicy.new_event_loop() states in its docstring:

The "policy registry" means settings the event loop as active for the policy in the thread local storage. 
The code is here:
https://github.com/python/cpython/blob/f235178beccf5eb5b47e770240f32d9ba24b26fd/Lib/asyncio/events.py#L659-L665

Reply:
> (I realize I got this backwards -- you are correctly calling set_event_loop() when you're calling new_event_loop(). But where is it written that if the user supplies a loop_factory function, that function must both create and set the event loop?)

A simple explanation is that if `asyncio` configure the event loop for you  then it will set the event loop in the policy registry so that you can get it as the current loop. However if you provide your own `loop_factory` then you need to manage setting the loop as current assuming in the registry aka thread local storage if you have one and manage child watchers yourself.